src/tls/httpredirect.c

```diff
@@ -0,0 +1,176 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config.h"
+
+#include "httpredirect.h"
+
+#include <sys/socket.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdbool.h>
+#include <string.h>
+#include <unistd.h>
+
+static char *
+read_line (FILE *stream,
+           char *buffer,
+           size_t sizeof_buffer,
+           size_t *offset)
+{
+  if (*offset >= sizeof_buffer)
+    return NULL;
+
+  char *line = fgets (buffer + *offset, sizeof_buffer - *offset, stream);
+
+  if (line == NULL)
+    return NULL;
+
+  /* Make sure we're terminated with \r\n or \n */
+  size_t line_length = strcspn (line, "\r\n");
+  const char *ending = line + line_length;
+  if (strcmp (ending, "\r\n") != 0 && strcmp (ending, "\n") != 0)
+    return NULL;
+
+  /* Discard the ending */
+  line[line_length++] = '\0';
+
+  *offset += line_length;
+
+  return line;
+}
+
+static bool
+write_error (FILE *output)
+{
+  fprintf (output, "HTTP/1.1 400 Client Error\r\n"
+                   "\r\n"
+                   "Incorrect request.\r\n");
+
+  return false;
+}
+
+static bool
+http_redirect (FILE *input,
+               FILE *output)
+{
+  char buffer[10000];
+  size_t offset = 0;
+
+  char *request_line = read_line (input, buffer, sizeof buffer, &offset);
+  if (request_line == NULL)
+    return write_error (output);
+
+  char *path = strchr (request_line, ' ');
```

Comment:
This entirely ignores the first word, i.e. the method (GET/PUT/POST etc.). Indeed that's not actually relevant for the behaviour here, and I think it's ok to send a redirect even for unsupported/bogus methods. I can't poke any holes in this, so let's keep it simple indeed.